### ISSUE TITLE
Performance: Allow plugin-transform-spread to use `Object.assign`

### ISSUE THREAD

**Issue by romgrk** [babel/babel#15673]
### 💻

- [X] Would you like to work on this feature?

### What problem are you trying to solve?

Hey, I've been doing a lot of performance profiling these last weeks (at [MUI](https://mui.com/)) and I've noted that [`_objectSpread2`](https://github.com/babel/babel/blob/02bacf882d624a9907f5b404f219ff74600d6d15/packages/babel-helpers/src/helpers/objectSpread2.js#L23) consistently is at the top of the list for most CPU cycles spent. Admittedly we're abusing the spread operator, but the problem is also that `_objectSpread2` is more than 10x slower than `Object.assign`, mainly because the spec [defines](https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#prod-SpreadElement) the spread operator in terms of `Object.defineProperty`, so it would be obviously incorrect to transform it to `Object.assign` by default.

_Chart: the `_objectSpread2` vs `Object.assign` vs spread benchmark results, for my system_
![image](https://github.com/babel/babel/assets/1423607/38a850c2-61b5-4f51-b136-19f7d237f7cf)

### Describe the solution you'd like

Although it's incorrect to transform the spread operator by default, in some cases it can make sense for developers to choose to do that transformation for their codebase, because they evaluate that it won't have negative effects.

So I would like to add an option to the `plugin-transform-operator` to enable it to use `Object.assign`.

### Describe alternatives you've considered

I'll publish a separate plugin if you're not interested, feel free to close this issue.

**Comment by nicolo-ribaudo** [comment]
You can enable the [setSpreadProperties](https://babeljs.io/docs/assumptions#setspreadproperties) assumption, and if the [targets](https://babeljs.io/docs/options#targets) you specify support `Object.assign` Babel will use that. If you don't specify the top-level targets option, with that assumption enabled Babel will fall back to the [`extends`](https://github.com/babel/babel/blob/02bacf882d624a9907f5b404f219ff74600d6d15/packages/babel-helpers/src/helpers.ts#L165) helper, which still uses `Object.assign` when available.

**Comment by nicolo-ribaudo** [comment]
Related - https://github.com/mui/material-ui/pull/37461

**Comment by romgrk** [comment]
Neat, thanks for the quick response. That assumption only has effects on the transform-spread plugin IIUC, right?

It would be nice to add a note about it in the docs, hard to find it otherwise.

**Comment by nicolo-ribaudo** [comment]
> That assumption only has effects on the transform-spread

It affects the `transform-object-rest-spread` and `transform-destructuring` plugin, since both are responsible of transforming object spread (the first one only runs when the second one isn't enabled).

> It would be nice to add a note about it in the docs, hard to find it otherwise.

mhh, we should probably have a "relevant assumptions" section in every plugin page.

**Comment by romgrk** [comment]
Alright. Great timing on that PR lol.

**Comment by JLHwung** [comment]
> It would be nice to add a note about it in the docs, hard to find it otherwise.

Yeah, the assumption was previously the loose mode: https://babel.dev/docs/babel-plugin-transform-object-rest-spread#loose. For every plugin with loose mode support, we provided the equivalent assumptions config.
